flocker/node/agents/ebs.py

```diff
@@ -511,30 +515,50 @@ def attach_volume(self, blockdevice_id, attach_to):
                 ebs_volume.status != 'available'):
             raise AlreadyAttachedVolume(blockdevice_id)
 
-        with self.lock:
-            # begin lock scope
-
-            blockdevices = FilePath(b"/sys/block").children()
-            device = self._next_device(attach_to)
-
-            if device is None:
-                # XXX: Handle lack of free devices in ``/dev/sd[f-p]`` range
-                # (see https://clusterhq.atlassian.net/browse/FLOC-1887).
-                # No point in attempting an ``attach_volume``, so, return.
-                return
-
-            self.connection.attach_volume(blockdevice_id, attach_to, device)
-
-            # Wait for new device to manifest in the OS. Since there
-            # is currently no standardized protocol across Linux guests
-            # in EC2 for mapping `device` to the name device driver picked (see
-            # http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/device_naming.html),
-            # let us wait for a new block device to be available to the OS,
-            # and interpret it as ours.
-            # Wait under lock scope to reduce false positives.
-            new_device = _wait_for_new_device(blockdevices, volume.size)
-
-            # end lock scope
+        attached = False
+        ignore_devices = pset([])
+        attach_attempts = 0
+        while (not attached and attach_attempts < MAX_RETRIES):
+            with self.lock:
+                # begin lock scope
+
+                blockdevices = FilePath(b"/sys/block").children()
+                device = self._next_device(attach_to, ignore_devices)
+
+                if device is None:
+                    # XXX: Handle lack of free devices in ``/dev/sd[f-p]``.
+                    # (https://clusterhq.atlassian.net/browse/FLOC-1887).
+                    # No point in attempting an ``attach_volume``, return.
+                    return
+
+                try:
+                    self.connection.attach_volume(blockdevice_id,
+                                                  attach_to,
+                                                  device)
+                except EC2ResponseError as e:
+                    # If attach failed because selected device is currently
+                    # in use, retry MAX_RETRIES times to find another
+                    # unused device.
+                    used = u"Attachment point {0} is already in use".format(
```

Comment:
This is brittle. Instead I suggest just retrying no matter what the message - worst case the convergence loop will try again if there's a real problem.